src/client/configuration.rs

```diff
@@ -27,33 +27,89 @@ use super::property_snapshot::PropertySnapshot;
 /// It contains a subset of models::ConfigurationJson, adding indexing.
 #[derive(Debug, Default)]
 pub(crate) struct Configuration {
-    pub(crate) features: HashMap<String, Feature>,
-    pub(crate) properties: HashMap<String, Property>,
-    pub(crate) segments: HashMap<String, Segment>,
+    features: HashMap<String, (Feature, SegmentRules)>,
+    properties: HashMap<String, (Property, SegmentRules)>,
 }
 
 impl Configuration {
+    /// Constructs the Configuration, by consuming and filtering data in exchange format
+    pub fn new(environment_id: &str, configuration: ConfigurationJson) -> Result<Self> {
+        let environment = configuration
+            .environments
+            .into_iter()
+            .find(|e| e.environment_id == environment_id)
+            .ok_or(ConfigurationAccessError::EnvironmentNotFound {
+                environment_id: environment_id.to_string(),
+            })?;
+        // FIXME: why not filtering for collection here?
+
+        let features = environment
+            .features
+            .into_iter()
+            .map(|mut feature| {
+                feature.segment_rules.sort_by(|a, b| a.order.cmp(&b.order));
+
+                // Get the segment rules that apply to this feature
+                let segments = Self::get_segments_for_segment_rules(
+                    &configuration.segments,
+                    &feature.segment_rules,
+                );
+
+                // Integrity DB check: all segment_ids should be available in the snapshot
+                if feature.segment_rules.len() != segments.len() {
+                    return Err(ConfigurationAccessError::MissingSegments {
+                        resource_id: feature.feature_id.to_string(),
+                    }
+                    .into());
+                }
+
+                let segment_rules =
+                    SegmentRules::new(segments, feature.segment_rules.clone(), feature.kind);
```

Comment:
We can compute the `SegmentRule` object associated to each `property`/`feature` and store it, instead of computing it every time we get it. 

I think `Configuration::new` is a good place to run _heavy_ work, it runs in the background after a HTTP call, so it should not interfere with application code.